x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/runtime/GeoShapeScriptFieldGeoShapeQuery.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.spatial.search.runtime;
+
+import org.apache.lucene.geo.Component2D;
+import org.apache.lucene.geo.LatLonGeometry;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.common.geo.Orientation;
+import org.elasticsearch.common.geo.ShapeRelation;
+import org.elasticsearch.geometry.Geometry;
+import org.elasticsearch.index.mapper.GeoShapeIndexer;
+import org.elasticsearch.script.GeometryFieldScript;
+import org.elasticsearch.script.Script;
+import org.elasticsearch.xpack.spatial.index.fielddata.CentroidCalculator;
+import org.elasticsearch.xpack.spatial.index.fielddata.Component2DVisitor;
+import org.elasticsearch.xpack.spatial.index.fielddata.CoordinateEncoder;
+import org.elasticsearch.xpack.spatial.index.fielddata.GeometryDocValueReader;
+import org.elasticsearch.xpack.spatial.index.fielddata.GeometryDocValueWriter;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.Arrays;
+import java.util.Objects;
+
+public class GeoShapeScriptFieldGeoShapeQuery extends AbstractGeoShapeScriptFieldQuery {
+
+    private final Component2DVisitor visitor;
+    private final LatLonGeometry[] geometries;
+    private final ShapeRelation relation;
+    private final GeoShapeIndexer indexer;
+    private final GeometryDocValueReader reader = new GeometryDocValueReader();
+
+    public GeoShapeScriptFieldGeoShapeQuery(
+        Script script,
+        GeometryFieldScript.LeafFactory leafFactory,
+        String fieldName,
+        ShapeRelation relation,
+        LatLonGeometry... geometries
+    ) {
+        super(script, leafFactory, fieldName);
+        this.geometries = geometries;
+        this.relation = relation;
+        final Component2D component2D = LatLonGeometry.create(geometries);
+        this.visitor = Component2DVisitor.getVisitor(component2D, relation.getLuceneRelation(), CoordinateEncoder.GEO);
+        indexer = new GeoShapeIndexer(Orientation.CCW, fieldName);
+    }
+
+    @Override
+    protected boolean matches(Geometry geometry) {
+        if (geometry == null) {
+            return false;
+        }
+        visitor.reset();
+        try {
+            reader.reset(encodeGeometry(geometry));
+            reader.visit(visitor);
```

Comment:
Is this going to be a problem with concurrency?  We have a single `reader` object here for the whole query but we may call `matches` from multiple segment tasks in parallel.

Reply:
good catch! I move the creation of the visitor to the matches. It will create one visitor for each call which is not great but compare to the cost of encoding the geometry, I don't think it is worthy to try to optimize it.

Reply:
The visitor is fine as a top-level field, it's the `reader` that needs to be done per-call I think?

Reply:
doh! yes the reader needs to be moved here but the visitor too, it holds a state when visited that is not thread safe.